File: ruby/bench.rb
# frozen_string_literal: true

# NXS vs JSON vs CSV benchmark — Ruby (stdlib only).
#
# 6 scenarios across 4 dataset sizes (1k, 10k, 100k, 1M records).
#
# Usage: ruby ruby/bench.rb [fixtures_dir]

require 'json'
require 'csv'
require_relative 'nxs'

# ── Helpers ──────────────────────────────────────────────────────────────────

# Returns average wall-clock seconds per iteration.
def bench(iters, &block)
  3.times(&block) # warmup
  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  iters.times(&block)
  (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) / iters
end

def fmt_time(s)
  return "#{(s * 1e9).round}  ns" if s < 1e-6
  return "#{format('%.1f', s * 1e6)} µs" if s < 1e-3
  return "#{format('%.2f', s * 1e3)} ms" if s < 1

  "#{'%.2f' % s}  s"
end

def fmt_bytes(n)
  return "#{n} B" if n < 1024
  return "#{format('%.1f', n / 1024.0)} KB" if n < 1024 * 1024

  "#{format('%.2f', n / 1024.0 / 1024.0)} MB"
end

def row(label, avg, baseline)
  ratio = if avg == baseline
            'baseline'
          elsif avg < baseline
            "#{format('%.1f', baseline / avg)}x faster"
          else
            "#{format('%.1f', avg / baseline)}x slower"
          end
  printf "  │  %-42s %12s   %s\n", label, fmt_time(avg), ratio
end

def section(title)
  bar = '─' * [0, 76 - title.length].max
  puts "\n  ┌─ #{title} #{bar}┐"
end

def endsection
  puts "  └#{'─' * 79}┘"
end

# ── Per-size benchmark ───────────────────────────────────────────────────────

def run_scale(fixture_dir, n)
  nxb_path  = File.join(fixture_dir, "records_#{n}.nxb")
  json_path = File.join(fixture_dir, "records_#{n}.json")
  csv_path  = File.join(fixture_dir, "records_#{n}.csv")

  unless File.exist?(nxb_path) && File.exist?(json_path)
    puts "\n  ⚠  skipping n=#{n}: fixtures missing"
    return
  end

  nxb_buf  = File.binread(nxb_path)
  json_str = File.read(json_path, encoding: 'UTF-8')
  csv_str  = File.exist?(csv_path) ? File.read(csv_path, encoding: 'UTF-8') : nil

  puts "\n#{'━' * 40}  n = #{n.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}  #{'━' * 38}"
  printf "  .nxb  size:  %10s\n", fmt_bytes(nxb_buf.bytesize)
  printf "  .json size:  %10s  (%.2fx NXS)\n", fmt_bytes(json_str.bytesize), json_str.bytesize.to_f / nxb_buf.bytesize

  # Scale iteration counts to keep wall-clock reasonable
  if n >= 1_000_000
    parse_iters = 3
    rand_iters = 5_000
    iter_iters = 2
    cold_iters = 3
  elsif n >= 100_000
    parse_iters = 20
    rand_iters = 20_000
    iter_iters = 5
    cold_iters = 15
  elsif n >= 10_000
    parse_iters = 200
    rand_iters = 50_000
    iter_iters = 50
    cold_iters = 100
  else
    parse_iters = 2_000
    rand_iters = 100_000
    iter_iters = 500
    cold_iters = 500
  end

  rng     = Random.new(0)
  indices = Array.new(rand_iters) { rng.rand(n) }

  # ── 1. Open file ────────────────────────────────────────────────────────────
  section('Open file (parse / index full structure)')
  t_json_open = bench(parse_iters) { JSON.parse(json_str) }
  row('JSON.parse(entire document)', t_json_open, t_json_open)

  if csv_str
    t_csv_open = bench(parse_iters) { CSV.parse(csv_str, headers: true) }
    row('CSV.parse(entire document)', t_csv_open, t_json_open)
  end

  t_nxs_open = bench(parse_iters) { Nxs::Reader.new(nxb_buf) }
  row('Nxs::Reader.new(buffer)', t_nxs_open, t_json_open)
  endsection

  # Pre-build warm objects for remaining benchmarks
  parsed  = JSON.parse(json_str)
  reader  = Nxs::Reader.new(nxb_buf)
  csv_tab = csv_str ? CSV.parse(csv_str, headers: true) : nil

  # ── 2. Warm random access ───────────────────────────────────────────────────
  section('Warm random-access read (1 field from 1 record)')
  ri = 0
  t_json_rand = bench(rand_iters) { parsed[indices[ri = (ri + 1) % rand_iters]]['username'] }
  row("parsed_json[k]['username']", t_json_rand, t_json_rand)

  if csv_tab
    ri2 = 0
    t_csv_rand = bench(rand_iters) { csv_tab[indices[ri2 = (ri2 + 1) % rand_iters]]['username'] }
    row("csv_table[k]['username']", t_csv_rand, t_json_rand)
  end

  ri3 = 0
  t_nxs_rand = bench(rand_iters) { reader.record(indices[ri3 = (ri3 + 1) % rand_iters]).get_str('username') }
  row("reader.record(k).get_str('username')", t_nxs_rand, t_json_rand)
  endsection

  # ── 3. Cold start ───────────────────────────────────────────────────────────
  section('Cold start — open file + read 1 field')
  cold_k = n / 2
  t_json_cold = bench(cold_iters) { JSON.parse(json_str)[cold_k]['username'] }
  row("JSON.parse + arr[k]['username']", t_json_cold, t_json_cold)

  if csv_str
    t_csv_cold = bench(cold_iters) { CSV.parse(csv_str, headers: true)[cold_k]['username'] }
    row("CSV.parse + table[k]['username']", t_csv_cold, t_json_cold)
  end

  t_nxs_cold = bench(cold_iters) { Nxs::Reader.new(nxb_buf).record(cold_k).get_str('username') }
  row('Nxs::Reader.new + record(k).get_str', t_nxs_cold, t_json_cold)
  endsection

  # ── 4. Full scan per-record ──────────────────────────────────────────────────
  section('Full scan per-record — sum score via record(i).get_f64')
  t_json_scan = bench(iter_iters) do
    s = 0.0
    parsed.each { |r| s += r['score'].to_f }
    s
  end
  row("arr.each { |r| sum += r['score'] }", t_json_scan, t_json_scan)

  if csv_tab
    t_csv_scan = bench(iter_iters) do
      s = 0.0
      csv_tab.each { |r| s += r['score'].to_f }
      s
    end
    row("csv.each { |r| sum += r['score'] }", t_csv_scan, t_json_scan)
  end

  t_nxs_scan = bench(iter_iters) do
    s = 0.0
    n.times do |i|
      v = reader.record(i).get_f64('score')
      s += v if v
    end
    s
  end
  row('n.times { reader.record(i).get_f64 }', t_nxs_scan, t_json_scan)
  endsection

  # ── 5. Reducer ──────────────────────────────────────────────────────────────
  section("Reducer — reader.sum_f64('score') tight loop")
  t_json_red = bench(iter_iters) do
    s = 0.0
    parsed.each { |r| s += r['score'].to_f }
    s
  end
  row('json: arr.each sum (baseline)', t_json_red, t_json_red)

  if csv_tab
    t_csv_red = bench(iter_iters) do
      s = 0.0
      csv_tab.each { |r| s += r['score'].to_f }
      s
    end
    row('csv: table.each sum', t_csv_red, t_json_red)
  end

  t_nxs_red = bench(iter_iters) { reader.sum_f64('score') }
  row("reader.sum_f64('score')", t_nxs_red, t_json_red)
  endsection

  # ── 6. Cold pipeline ────────────────────────────────────────────────────────
  section('Cold pipeline — read file from disk + sum scores')
  t_json_pipe = bench(cold_iters) do
    s = 0.0
    JSON.parse(json_str).each { |r| s += r['score'].to_f }
    s
  end
  row('JSON.parse + sum scores', t_json_pipe, t_json_pipe)

  t_nxs_pipe = bench(cold_iters) { Nxs::Reader.new(nxb_buf).sum_f64('score') }
  row('Nxs::Reader.new + sum_f64', t_nxs_pipe, t_json_pipe)
  endsection
end

# ── Main ─────────────────────────────────────────────────────────────────────

fixture_dir = ARGV[0] || '../js/fixtures'

puts
puts '╔════════════════════════════════════════════════════════════════════════════════╗'
puts '║         NXS vs JSON vs CSV  —  Ruby Benchmark                                 ║'
puts '╚════════════════════════════════════════════════════════════════════════════════╝'
printf "\n  Ruby:     %s\n", RUBY_DESCRIPTION
printf "  Platform: %s-%s\n", RUBY_PLATFORM, RbConfig::CONFIG['host_cpu']
printf "  Fixtures: %s\n", fixture_dir

[1_000, 10_000, 100_000, 1_000_000].each do |n|
  run_scale(fixture_dir, n)
rescue StandardError => e
  puts "\n  ⚠  n=#{n} failed: #{e.message}"
  puts(e.backtrace.first(3).map { |l| "       #{l}" })
end

puts
puts '═' * 80
puts '  Notes:'
puts '    • JSON.parse materialises the entire document eagerly (Ruby stdlib C extension).'
puts '    • Nxs::Reader reads only preamble + schema + tail-index on open (~microseconds).'
puts '    • .record(k) is O(1) — one 10-byte tail-index read.'
puts '    • sum_f64 is an allocation-free tight loop over the bitmask + 8-byte value.'
puts '    • CSV is included as a second baseline; all times are per-iteration averages.'
puts

File: py/bench.py
"""NXS vs JSON benchmark — Python (stdlib only).

Compares:
  1. Open / parse full structure
  2. Random access — 1 field from record k (warm)
  3. Cold start — open file + read 1 field
  4. Full scan — sum of 'score' across all records

Usage: python3 bench.py [fixtures_dir]
"""
from __future__ import annotations

import json
import random
import sys
import time
from pathlib import Path

from nxs import NxsReader


def bench(iters: int, fn) -> float:
    """Return average seconds per iteration."""
    for _ in range(max(3, iters // 10)):
        fn()  # warmup
    start = time.perf_counter()
    for _ in range(iters):
        fn()
    return (time.perf_counter() - start) / iters


def fmt_time(s: float) -> str:
    if s < 1e-6:   return f"{s * 1e9:.0f} ns"
    if s < 1e-3:   return f"{s * 1e6:.1f} µs"
    if s < 1:      return f"{s * 1e3:.2f} ms"
    return f"{s:.2f} s"


def fmt_bytes(n: int) -> str:
    if n < 1024:            return f"{n} B"
    if n < 1024 * 1024:     return f"{n/1024:.1f} KB"
    return f"{n/1024/1024:.2f} MB"


def row(label: str, avg: float, baseline: float) -> None:
    if baseline == avg:
        ratio = "baseline"
    elif avg < baseline:
        ratio = f"{baseline/avg:.1f}x faster"
    else:
        ratio = f"{avg/baseline:.1f}x slower"
    print(f"  │  {label:<42} {fmt_time(avg):>10}   {ratio}")


def section(title: str) -> None:
    print(f"\n  ┌─ {title} {'─' * max(0, 76 - len(title))}┐")


def endsection() -> None:
    print(f"  └{'─' * 79}┘")


def run_scale(fixture_dir: Path, n: int) -> None:
    nxb_path = fixture_dir / f"records_{n}.nxb"
    json_path = fixture_dir / f"records_{n}.json"
    if not nxb_path.exists() or not json_path.exists():
        print(f"\n  ⚠  skipping n={n}: fixtures missing")
        return

    nxb_buf = nxb_path.read_bytes()
    json_str = json_path.read_text()

    print(f"\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  n = {n:,}  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"  .nxb size:  {fmt_bytes(len(nxb_buf)):>10}")
    print(f"  .json size: {fmt_bytes(len(json_str)):>10}  ({len(json_str)/len(nxb_buf):.2f}x NXS)")

    # Iteration counts scaled to wall-clock budget
    if n >= 1_000_000:
        parse_iters, random_iters, iterate_iters, cold_iters = 5, 20_000, 3, 5
    elif n >= 100_000:
        parse_iters, random_iters, iterate_iters, cold_iters = 30, 50_000, 10, 30
    elif n >= 10_000:
        parse_iters, random_iters, iterate_iters, cold_iters = 300, 100_000, 100, 200
    else:
        parse_iters, random_iters, iterate_iters, cold_iters = 3000, 100_000, 1000, 1000

    # ── 1. Open / Parse ─────────────────────────────────────────────────────
    section("Open file (parse full structure)")
    t_json_open = bench(parse_iters, lambda: json.loads(json_str))
    row("json.loads(entire document)", t_json_open, t_json_open)
    t_nxs_open = bench(parse_iters, lambda: NxsReader(nxb_buf))
    row("NxsReader(buffer)", t_nxs_open, t_json_open)
    endsection()

    # ── 2. Warm random access ───────────────────────────────────────────────
    section("Random-access read (1 field from 1 record, averaged over random k)")

    parsed = json.loads(json_str)
    reader = NxsReader(nxb_buf)

    rng = random.Random(0)
    indices = [rng.randrange(n) for _ in range(random_iters)]
    it = iter(indices)

    # Pre-materialize iterator to avoid StopIteration overhead in loop
    def json_lookup():
        k = next(it)
        return parsed[k]["username"]

    t_json_random = bench(random_iters, lambda: parsed[indices[rng.randrange(random_iters)]]["username"])
    row("arr[k]['username'] (pre-parsed)", t_json_random, t_json_random)

    t_nxs_random = bench(random_iters, lambda: reader.record(indices[rng.randrange(random_iters)]).get_str("username"))
    row("reader.record(k).get_str('username')", t_nxs_random, t_json_random)
    endsection()

    # ── 3. Cold start ───────────────────────────────────────────────────────
    section("First access — open file + read 1 field (cold start)")
    cold_idx = n // 2
    t_json_cold = bench(cold_iters, lambda: json.loads(json_str)[cold_idx]["username"])
    row("json.loads + arr[k]['username']", t_json_cold, t_json_cold)
    t_nxs_cold = bench(cold_iters, lambda: NxsReader(nxb_buf).record(cold_idx).get_str("username"))
    row("NxsReader + record(k).get_str(...)", t_nxs_cold, t_json_cold)
    endsection()

    # ── 4. Full scan ────────────────────────────────────────────────────────
    section("Full scan — sum of 'score' field across all records")

    def json_scan():
        s = 0.0
        for r in parsed:
            s += r["score"]
        return s

    def nxs_scan():
        s = 0.0
        for r in reader.records():
            s += r.get_f64("score")
        return s

    t_json_scan = bench(iterate_iters, json_scan)
    row("for r in arr: sum += r['score']", t_json_scan, t_json_scan)
    t_nxs_scan = bench(iterate_iters, nxs_scan)
    row("for r in reader.records(): sum += ...", t_nxs_scan, t_json_scan)
    endsection()


def main() -> int:
    fixture_dir = Path(sys.argv[1] if len(sys.argv) > 1 else "../js/fixtures")

    print("\n╔════════════════════════════════════════════════════════════════════════════════╗")
    print("║           NXS vs JSON  —  Python Benchmark                                    ║")
    print("╚════════════════════════════════════════════════════════════════════════════════╝")
    print(f"\n  Python:   {sys.version.split()[0]}")
    print(f"  Platform: {sys.platform}")
    print(f"  Fixtures: {fixture_dir}")

    for n in (1_000, 10_000, 100_000, 1_000_000):
        try:
            run_scale(fixture_dir, n)
        except Exception as e:
            print(f"\n  ⚠  n={n} failed: {e}")

    print("\n" + "═" * 80)
    print("  Notes:")
    print("    • json.loads parses the entire document eagerly.")
    print("    • NxsReader reads only the preamble + schema + tail-index header.")
    print("    • .record(k) is O(1) — no scan through earlier records.")
    print("    • Python's json is written in C; NXS reader is pure Python (struct/memoryview).\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
